src/server/notifications/StreamingHttpChannel2023/StreamingHttp2023Emitter.ts

```diff
@@ -0,0 +1,38 @@
+import { getLoggerFor } from '../../../logging/LogUtil';
+
+import type { Representation } from '../../../http/representation/Representation';
+import { AsyncHandler } from '../../../util/handlers/AsyncHandler';
+import type { NotificationChannel } from '../NotificationChannel';
+import type { StreamingHttpMap } from './StreamingHttpMap';
+
+export interface StreamingHttpEmitterInput {
+  channel: NotificationChannel;
+  representation: Representation;
+}
+
+/**
+ * Emits notifications on StreamingHTTPChannel2023 streams.
+ * Uses the response streams found in the provided map.
+ * The key should be the identifier of the topic resource.
+ */
+export class StreamingHttp2023Emitter extends AsyncHandler<StreamingHttpEmitterInput> {
+  protected readonly logger = getLoggerFor(this);
+
+  public constructor(
+    private readonly streamMap: StreamingHttpMap,
+  ) {
+    super();
+  }
+
+  public async handle({ channel, representation }: StreamingHttpEmitterInput): Promise<void> {
+    // Called as a NotificationEmitter: emit the notification
+    const streams = this.streamMap.get(channel.topic);
+    if (streams) {
+      for (const stream of streams) {
+        representation.data.pipe(stream, { end: false });
+      }
+    } else {
+      representation.data.destroy();
```

Comment:
When I start the server, change gets emitted for `http://localhost:3000/.internal/setup/current-server-version`, which leads to this line resulting in `[GuardedStream] {Primary} error: No error listener was attached but error was thrown: premature close`
I'm considering changing the `StreamingHttpListeningActivityHandler` not to emit notifications for changes on resources matching `./internal/`

Reply:
I would expect this error to then get thrown every time a resource is modified and `streams` is not defined, is that not the case?

What happens if you first attach an error listener to the data stream before destroying. Does it still give an error then? It's a safety mechanism in CSS to make sure there is always an error listener attached to a stream.

Reply:
> I would expect this error to then get thrown every time a resource is modified and streams is not defined, is that not the case?

You are right!

Since the WebSocket2023Emmiter does the same https://github.com/CommunitySolidServer/CommunitySolidServer/blob/f59b2c29704f48c02bcae13787588322524b2511/src/server/notifications/WebSocketChannel2023/WebSocket2023Emitter.ts#L33

I ran a quick test, and no error was logged if I created a web socket channel without specifying `accept`.
On the other hand, the same error was logged if I created a web socket channel while specifying `accept: text/turtle`. I haven't looked into how representations get converted, but it seems like a stream without a listener might be created in that process.
I'm going to triple-check it, but if the same error gets logged for WebSocketChannel2023 with `accept: text/turtle`, can we track it as a separate issue and solve it outside of this PR?

-----

I still want to double-check with you if it is intentional that the `ResourceStore` emits changes on resources inside `.internal` directory. If it is, do you see it reasonable if I filter using a regex to check the resource identifier and don't emit notifications for anything inside the `.internal`. I'm also wondering if I will need to pass the `baseUrl` to my `StreamingHttpListeningActivityHandler` or could use a regex without the `baseUrl`?

Reply:
If I understand correctly, the difference is that with the WebSocket solution, that code will only be reached if someone registered a channel to listen to that resource. Otherwise no notification would be created and the server would not attempt to emit it.

The difference here, and this is related to the second question as well, is that `StreamingHttpListeningActivityHandler` will always generate a notification and always attempt to emit it, as it doesn't know if someone is listening or not, causing it to also generate notifications for the `.internal` container, and any other resource that gets modified but nobody is listening to. With the other notification solutions this gets cut off before the notification gets generated instead. As a solution, it might make sense to also give the `StreamingHttpListeningActivityHandler` access to the `StreamingHttpMap` so it can do an existence check before calling the `NotificationHandler`?